File: backend/routing/database/update.go
package routes_db

import (
	"aura/database"
	"aura/logging"
	"aura/models"
	"aura/utils/httpx"
	"net/http"
)

type updateItemRequest struct {
	Complete   bool               `json:"complete"`
	UpdateItem models.DBSavedItem `json:"update_item"`
}

type updateItemResponse struct {
	Result string `json:"result"`
}

// UpdateItemInDB godoc
// @Summary      Update Item In Database
// @Description  Update a Media Item and its associated Poster Sets in the database. Poster Sets marked with "to_delete" will be removed, while others will be upserted.
// @Tags         Database
// @Accept       json
// @Produce      json
// @Param        req  body      updateItemRequest  true  "Update Item Request"
// @Security 	 BearerAuth
// @Failure      401  {object}  httpx.UnauthorizedResponse "Unauthorized (only when Auth.Enabled=true)"
// @Success      200           {object}  httpx.JSONResponse{data=updateItemResponse}
// @Failure      500           {object}  httpx.JSONResponse "Internal Server Error"
// @Router       /api/db [patch]
func UpdateItemInDB(w http.ResponseWriter, r *http.Request) {
	ctx, ld := logging.CreateLoggingContext(r.Context(), r.URL.Path)
	logAction := ld.AddAction("Update Item In Database", logging.LevelInfo)
	ctx = logging.WithCurrentAction(ctx, logAction)

	var req updateItemRequest
	var response updateItemResponse

	// Parse the request body

	Err := httpx.DecodeRequestBodyToJSON(ctx, r.Body, &req, "Update Item - Decode Request Body")
	if Err.Message != "" {
		httpx.SendResponse(w, ld, response)
		return
	}
	logAction.AppendResult("complete", req.Complete)

	for _, ps := range req.UpdateItem.PosterSets {
		if ps.ToDelete {
			// Delete the poster set
			Err := database.DeletePosterSetForMediaItem(ctx, req.UpdateItem.MediaItem.TMDB_ID, req.UpdateItem.MediaItem.LibraryTitle, ps.ID)
			if Err.Message != "" {
				httpx.SendResponse(w, ld, response)
				return
			}
		} else {
			// Upsert the poster set
			Err := database.UpsertSavedItem(ctx, req.UpdateItem)
			if Err.Message != "" {
				httpx.SendResponse(w, ld, response)
				return
			}
		}
	}

	response.Result = "ok"
	httpx.SendResponse(w, ld, response)
}

File: frontend/src/services/database/update.ts
import apiClient from "@/services/api-client";
import { ReturnErrorMessage } from "@/services/api-error-return";

import { log } from "@/lib/logger";
import { useLibrarySectionsStore } from "@/lib/stores/global-store-library-sections";

import type { APIResponse } from "@/types/api/api-response";
import type { DBSavedItem } from "@/types/database/db-poster-set";

export interface UpdateItem_Request {
  complete: boolean;
  update_item: DBSavedItem;
}

export interface UpdateItem_Response {
  result: string;
}

export const UpdateItemInDB = async (saveItem: DBSavedItem): Promise<APIResponse<UpdateItem_Response>> => {
  log(
    "INFO",
    "API - DB",
    "Update",
    `Patching ${saveItem.media_item.title} (${saveItem.media_item.tmdb_id} | ${saveItem.media_item.library_title}) in DB`,
    saveItem
  );
  for (const set of saveItem.poster_sets) {
    set.last_downloaded = new Date().toISOString();
  }

  try {
    const req: UpdateItem_Request = {
      update_item: saveItem,
      complete: true,
    };
    const response = await apiClient.patch<APIResponse<UpdateItem_Response>>(`/db`, req);
    if (response.data.status === "error") {
      throw new Error(response.data.error?.message || "Unknown error patching item in DB");
    } else {
      log(
        "INFO",
        "API - DB",
        "Update",
        `Patched ${saveItem.media_item.title} (${saveItem.media_item.tmdb_id} | ${saveItem.media_item.library_title}) in DB`,
        response.data
      );
    }
    const { upsertMediaItemSavedSet } = useLibrarySectionsStore.getState();
    for (const savedSet of saveItem.poster_sets) {
      upsertMediaItemSavedSet({
        tmdbID: saveItem.media_item.tmdb_id,
        libraryTitle: saveItem.media_item.library_title,
        setID: savedSet.id,
        setUser: savedSet.user_created,
        selectedTypes: savedSet.selected_types,
        toDelete: savedSet.to_delete,
      });
    }

    return response.data;
  } catch (error) {
    log(
      "ERROR",
      "API - DB",
      "Update",
      `Failed to patch ${saveItem.media_item.title} (${saveItem.media_item.tmdb_id} | ${saveItem.media_item.library_title}) in DB: ${
        error instanceof Error ? error.message : "Unknown error"
      }`,
      error
    );
    return ReturnErrorMessage<UpdateItem_Response>(error);
  }
};
